### PR TITLE
feat(logging): Add PyTorch-Native Deep Integration & Pure Logging Bri…

### LINKED ISSUE #1205: [Feature Request] PyTorch-Native Deep Integration & Pure Logging Bridge for Visdom
**Is your feature request related to a problem? Please describe.**
Currently users have to manually write repetitive `vis.line()` calls for every metric (loss, accuracy, learning rate, etc.) in every training loop. This boilerplate makes Visdom much less convenient than other tools. There is also no official, clean integration with PyTorch Lightning, so Lightning users either duplicate the same manual logging code or switch to heavier solutions.

**Describe the solution you'd like**
Add a thin, pure logging bridge that respects Visdom’s role as a visualization backend only:

- `VisdomLogger`: a Lightning `Logger` subclass so users can simply do `Trainer(logger=VisdomLogger(...))` and all `self.log(name, value)` calls are automatically routed to the correct `viz.line()` windows.
- A simple context manager / decorator for native PyTorch training loops.
- Automatic detection and logging of common metrics.
- Smart multi-run environment versioning.
- Configurable opt-in / opt-out.

No gradient computation, no autograd hooks, and no training logic will live inside Visdom. Users compute any extra values (e.g. gradient norms) themselves and pass them to the logger.
This matches the design philosophy already written in my GSoC proposal.

**Describe alternatives you've considered**

- Keep using manual `vis.line()` calls (current situation — very repetitive).
- Use TensorBoard, Weights & Biases or MLflow (these work well but add cloud dependencies, accounts, or heavy setup, which defeats Visdom’s lightweight local-first advantage).

**Additional context**

- PR #1003  was closed by mentor @marcoag because it included gradient hooks and computation logic (outside Visdom’s scope). I fully agree with the feedback and will open a completely clean PR once this issue is discussed.
- I have already completed all starter tasks and delivered several small cleanup PRs (#966, #1099, #1105).
- Happy to adjust the exact API based on mentor feedback before opening the new PR.

Looking forward to your thoughts!
CC: @marcoag @norbusan @mariobehling

### PR DESCRIPTION
**Pull request by pranayr710**
…dge (closes #1205)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code refactor or cleanup (changes to existing code for improved readability or performance)

## Checklist:

- [x] I adapted the version number under `py/visdom/VERSION` according to [Semantic Versioning](https://semver.org/)
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

[Feature Request] PyTorch-Native Deep Integration & Pure Logging Bridge for Visdom

## Description

Adds a lightweight, pure logging bridge (`visdom.logging`) to eliminate repetitive `vis.line()` boilerplate in PyTorch and PyTorch Lightning training loops. 

This PR introduces:
1. `VisdomLoggingHandler` — A context manager and decorator for vanilla PyTorch loops that manages window IDs, updates plots sequentially via `append`, and supports logging text and images.
2. `VisdomLogger` — A PyTorch Lightning `Logger` subclass that integrates seamlessly with Lightning's `Trainer` API (logs metrics to `viz.line` and hyperparams/status to `viz.text`).
3. Smart Environment Auto-versioning — Automatically increments experiment environments across runs (e.g. `run_000`, `run_001`).
4. Metric Filtering — Allows users to include/exclude logged metrics using glob patterns.
5. Verification Suite — Comprehensive mock tests and a working demo training script.

## Motivation and Context

Closes #1205.

Currently, logging multiple metrics in PyTorch requires writing and maintaining multiple `vis.line()` calls, tracking window IDs, and specifying coordinates manually. This boilerplate makes Visdom tedious to use compared to other modern loggers.

This logging bridge simplifies logging while respecting Visdom's design philosophy as a pure visualization backend: it contains **no gradient computation, no autograd hooks, and no model training logic** (addressing the feedback that led to PR #1003 being closed). PyTorch Lightning is kept as a completely optional dependency.

## How Has This Been Tested?

1. **Unit Tests**: Added 27 unit tests in `test/test_logging.py` covering auto-versioning, metric filtering, context managers/decorators, thread safety, and Lightning integrations.
2. **Regression Check**: Ran the existing test suite (`test_caption.py`) to verify zero regressions.
3. **Demo Script**: Created a runnable demonstration at `example/demo_logging.py` that exercises the features against a live server.

### Proof of Working (Test Execution):
```bash
python -m pytest test/ -v


============================= test session starts =============================
platform win32 -- Python 3.10.10, pytest-9.0.3, pluggy-1.6.0
rootdir: C:\Users\himes\pranay\fossasia\visdom
plugins: anyio-4.13.0, asyncio-1.3.0, cov-7.1.0
collected 35 items

test/test_caption.py::TestAudioCaption::test_caption_appears_in_html PASSED [  2%]
test/test_caption.py::TestAudioCaption::test_caption_is_html_escaped PASSED [  5%]
test/test_caption.py::TestAudioCaption::test_empty_caption_leaves_html_unchanged PASSED [  8%]
test/test_caption.py::TestAudioCaption::test_no_caption_leaves_html_unchanged PASSED [ 11%]
test/test_caption.py::TestVideoCaption::test_caption_appears_in_html PASSED [ 14%]
test/test_caption.py::TestVideoCaption::test_caption_is_html_escaped PASSED [ 17%]
test/test_caption.py::TestVideoCaption::test_empty_caption_leaves_html_unchanged PASSED [ 20%]
test/test_caption.py::TestVideoCaption::test_no_caption_leaves_html_unchanged PASSED [ 22%]
test/test_logging.py::TestGetNextVersion::test_custom_base PASSED        [ 25%]
test/test_logging.py::TestGetNextVersion::test_empty_env_list PASSED     [ 28%]
test/test_logging.py::TestGetNextVersion::test_gap_ignored PASSED        [ 31%]
test/test_logging.py::TestGetNextVersion::test_non_matching_envs_ignored PASSED [ 34%]
test/test_logging.py::TestGetNextVersion::test_sequential PASSED         [ 37%]
test/test_logging.py::TestGetNextVersion::test_server_unreachable PASSED [ 40%]
test/test_logging.py::TestVisdomLoggingHandler::test_context_manager PASSED [ 42%]
test/test_logging.py::TestVisdomLoggingHandler::test_creates_window_once PASSED [ 45%]
test/test_logging.py::TestVisdomLoggingHandler::test_decorator PASSED    [ 48%]
test/test_logging.py::TestVisdomLoggingHandler::test_env_property PASSED [ 51%]
test/test_logging.py::TestVisdomLoggingHandler::test_exclude_filter PASSED [ 54%]
test/test_logging.py::TestVisdomLoggingHandler::test_include_filter PASSED [ 57%]
test/test_logging.py::TestVisdomLoggingHandler::test_log_image PASSED    [ 60%]
test/test_logging.py::TestVisdomLoggingHandler::test_log_routes_to_line PASSED [ 62%]
test/test_logging.py::TestVisdomLoggingHandler::test_log_text PASSED     [ 65%]
test/test_logging.py::TestVisdomLoggingHandler::test_multiple_metrics_get_separate_windows PASSED [ 68%]
test/test_logging.py::TestVisdomLoggingHandler::test_step_none PASSED    [ 71%]
test/test_logging.py::TestVisdomLogger::test_exclude_filter PASSED       [ 74%]
test/test_logging.py::TestVisdomLogger::test_experiment_property PASSED  [ 77%]
test/test_logging.py::TestVisdomLogger::test_finalize PASSED             [ 80%]
test/test_logging.py::TestVisdomLogger::test_implements_interface PASSED [ 82%]
test/test_logging.py::TestVisdomLogger::test_include_filter PASSED       [ 85%]
test/test_logging.py::TestVisdomLogger::test_log_hyperparams PASSED      [ 88%]
test/test_logging.py::TestVisdomLogger::test_log_hyperparams_namespace PASSED [ 91%]
test/test_logging.py::TestVisdomLogger::test_log_metrics PASSED          [ 94%]
test/test_logging.py::TestVisdomLogger::test_log_metrics_with_global_step_key PASSED [ 97%]
test/test_logging.py::TestVisdomLogger::test_name_and_version PASSED     [100%]

============================= 35 passed in 27.50s =============================

### CODE REVIEW

**Comment by vedansh-5** [comment]
Hi @pranayr710, thanks for your contribution. 
This PR implements a large addition to the current codebase. We do not accept large changes from new contributors. Please take up a beginner friendly issue from the list of issues. We decided to close this PR.
Thanks~

**Comment by pranayr710** [comment]
Hi Vedansh,

Thank you for letting me know. I understand your policy regarding large
contributions from new members. I will look through the list of
beginner-friendly issues and find a smaller task to take up instead.

Best regards,

Raparla Meghapranay
